src/pages/Wallet/TwoFactorModal.tsx

```diff
@@ -0,0 +1,164 @@
+import React, {useEffect, useState} from "react";
+import Box from "@mui/material/Box";
+import Button from "@mui/material/Button";
+import {useExtensionStyles} from "../../styles/extension.styles";
+import {Modal} from "@unstoppabledomains/ui-components";
+import ManageInput from "@unstoppabledomains/ui-components/components/Manage/common/ManageInput";
+import useFireblocksAccessToken from "@unstoppabledomains/ui-components/hooks/useFireblocksAccessToken";
+import {setWalletPreferences} from "../../lib/wallet/preferences";
+import {WalletPreferences} from "../../types/wallet/preferences";
+import {
+  disableTwoFactor,
+  getTwoFactorChallenge,
+  getTwoFactorStatus,
+  verifyTwoFactorChallenge,
+} from "../../lib/wallet/2fa";
+import CircularProgress from "@mui/material/CircularProgress";
+import {QRCode} from "react-qrcode-logo";
+import {PreferenceSection} from "./Preferences";
+import {StorageSyncKey, chromeStorageGet} from "../../lib/chromeStorage";
+
+interface TwoFactorModalProps {
+  open?: boolean;
+  onClose: () => void;
+  preferences: WalletPreferences;
+  setPreferences: (v: WalletPreferences) => void;
+}
+
+export const TwoFactorModal: React.FC<TwoFactorModalProps> = ({
+  open,
+  onClose,
+  preferences,
+  setPreferences,
+}) => {
+  const {classes, cx} = useExtensionStyles();
+  const getAccessToken = useFireblocksAccessToken();
+  const [accessToken, setAccessToken] = useState("");
+  const [emailAddress, setEmailAddress] = useState<string>();
+  const [qrCodeContent, setQrCodeContent] = useState<string>();
+  const [errorMessage, setErrorMessage] = useState<string>();
+  const [otp, setOtp] = useState<string>();
+
+  useEffect(() => {
+    const loadAccessToken = async () => {
+      setAccessToken(await getAccessToken());
+      setEmailAddress(await chromeStorageGet(StorageSyncKey.Account));
+    };
+    void loadAccessToken();
+  }, []);
+
+  useEffect(() => {
+    if (!accessToken || !emailAddress) {
+      return;
+    }
+    const loadStatus = async () => {
+      // retrieve status
+      const enabled = await getTwoFactorStatus(accessToken);
+      updateStatus(enabled);
+
+      // if not enabled, retrieve QR code data
+      if (!enabled) {
+        const challenge = await getTwoFactorChallenge(accessToken);
+        const encodedChallenge = `otpauth://totp/${emailAddress}?secret=${challenge}`;
```

Comment:
We should have "Unstoppable Light Wallet" or something in here too

https://github.com/google/google-authenticator/wiki/Key-Uri-Format

Reply:
nice, yeah looks like we could add `&issuer=Unstoppable%20Lite%20Wallet` to the end of the encoded data.

Reply:
Followup, attempted to add `issuer` as both a query parameter and an encoded prefix to the user (e.g. `name:user`). Both methods fail to register in the OTP apps as a code when scanning the QR code. So I think we should leave this out of the encoding.

Reply:
you might need to `encodeURIComponent(...)` each of the things you are putting into this uri

Reply:
good idea, updated 👍